Sharc/Sharc/Sharc.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 20/09/13               \" DATE 
.Dt Sharc 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm Sharc,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1 description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner

CHANGELOG.md
0.9.9 beta
----------
<i>August 26, 2013</i>
* Added maximum dictionary reset cycle to ensure possible parallelization
* Improved XOR mask
* Reorganized global variables
* Added buffer size and dictionary max cycles in stream header
* Reduced dictionary sizes to optimize fitting in processor caches
* Removed "cipher" references in file names, functions, and preprocessor macro guard names
* Added dictionary preloading to optimize encoding/decoding startup efficiency

0.9.8 beta
----------
<i>August 19, 2013</i>
* Added preprocessor guards on large file #defines
* Modified dictionary indexing process to enable reuse of long distance previously-read file data (major performance boost)
* Namespaced dictionaries
* Added the specify output path (-p) option
* Fixed problem with some strings not being initialized in client.c
* Added parenthesis protection on preprocessor macro SHARC_PREFERRED_BUFFER_SIZE
* Added reset dictionary information in block header
* Modified -std=c99 to -std=gnu99 in Makefile to remove the ftello implicit declaration warning

0.9.7 beta
----------
<i>August 8, 2013</i>
* Improved binary mode setting for stdin/stdout on Windows
* Added support for big endian CPU systems
* Date/time display now using preprocessor macros
* Namespacing of macros, structures, types and functions names
* Modified platform detection compiler check strings
* Client now displaying endianness
* Modification of the file format to include the type of media (stream, file...)
* Changed file type identification check to a magic number
* Improved header functions signatures
* Cleared remaining compilation warnings
* Modified XOR hash mask

0.9.6 beta
----------
<i>August 1, 2013</i>
* Target architecture definition option re-added in makefile
* Improved makefile and program displays
* Now properly using stdin/stdout as binary streams on Windows
* Added support for very large files (> 4GB)
* Platform detection now using gcc preprocessor macros
* Fixed problem with -c1 compression not properly working with certain files

0.9.5 beta
----------
<i>July 20, 2013</i>
* Improved makefile and added build information on version display
* Added stddef.h to globals.h for standard macro definitions

0.9.4 beta
----------
<i>July 5, 2013</i>
* Added no prompting for overwrites option
* Created client interface to differentiate file management and stream compression
* Added stdin reading / stdout writing support
* Added help option to display usage

0.9.3 beta
----------
<i>July 2, 2013</i>
* Corrected out file name problem
* Corrected problem with intermediate buffer pointer swap in case of an encoding mode reversion
* Automatic generation of objects from sources list in makefile
* Time is now measured and displayed in seconds
* Modified timing function to provide better precision on Linux platforms

0.9.2 beta
----------
<i>June 29, 2013</i>
* Improved makefile
* Added usage
* Added check for file overwrites
* Corrected dual pass compression / decompression
* Updated license
* Added restoring of original file attributes
* Corrected problems with intermediate buffering

0.9.1 beta
----------
<i>June 24, 2013</i>  
* Initial beta version


src/Sharc.cpp
/*
 * Copyright (c) 2013, Centaurean
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of Centaurean nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL Centaurean BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * acceLZW
 *
 * 11/05/13 18:19
 * @author gpnuma
 */

#include "Sharcz.h"

Sharc::Sharc() {
    //dictionary = new ENTRY[1 << HASH_BITS];
	reset();
}

Sharc::~Sharc() {
    //delete[] dictionary;
}

FORCE_INLINE static void computeHash(unsigned int* hash, const unsigned int* value) {
    *hash = HASH_OFFSET_BASIS;
    *hash ^= *value;
    *hash *= HASH_PRIME;
    *hash = (*hash >> (32 - HASH_BITS)) ^ (*hash & 0xFFFF);
}

FORCE_INLINE static bool updateEntry(ENTRY* entry, const unsigned int* buffer, const unsigned int* index, SharcWriter* writer) {
	//entry->offset = *index;
	//entry->exists = true;
//#if __BYTE_ORDER == __LITTLE_ENDIAN
	//*(unsigned int*)entry = (*index | 1 << 24);
	/*entry->offset_0 = *index & 0xFFFF;
	entry->offset_1 = *index >> 16;
	entry->exists = true;*/
	*(unsigned int*)entry = (*index & 0xFFFFFF) | (1 << 24);
	//std::bitset<32> c(*(unsigned int*)entry);
	//std::bitset<32> i(*index);
    ////*(unsigned int*)entry = (*index << 1 | 1);
	////std::cout << sizeof(entry) << std::endl;
	//std::bitset<32> x(*index << 1 | 1);
	////std::bitset<32> y(*(unsigned int*)entry);
	//std::cout << sizeof(entry) /*<< ", " << c << ", " << i << " = " << x*/ << std::endl;
//#elif __BYTE_ORDER == __BIG_ENDIAN
//    *(unsigned int*)entry = (*index << 1 | 1);
//#elif
//#error
//#endif
	//writer->writeMode(false);
	return writer->writeChunk((unsigned int)*(buffer + *index));
}

FORCE_INLINE bool Sharc::compress(byte* byteBuffer, SharcWriter* writer) {
    //unsigned int outBits = 0;
    unsigned int hash;
    
    const unsigned int length = writer->getLimit() >> 2;
    const unsigned int* buffer = (const unsigned int*)byteBuffer;
    
    for(unsigned int i = 0; i < length; i ++) {
        computeHash(&hash, buffer + i);
        ENTRY* found = &dictionary[hash];
        if(found->exists) {
			if(buffer[i] ^ buffer[*(unsigned int*)found & 0xFFFFFF/*found->offset_0 + (found->offset_1 << 16)*/]) {
                if(!updateEntry(found, buffer, &i, writer))
                    return false;
            } else {
				//writer->writeMode(true);
                if(!writer->writeChunk((unsigned short)hash))
                    return false;
                //outBits += (HASH_BITS + 1);
			}
        } else {
            if(!updateEntry(found, buffer, &i, writer))
                return false;
        }
    }
    
	//return outBits;
    return true;
}

FORCE_INLINE bool Sharc::subs(byte* byteBuffer, SubsWriter* writer) {
    const unsigned int length = writer->getLimit() >> 2;
    const unsigned int* buffer = (const unsigned int*)byteBuffer;
    
    for(unsigned int i = 0; i < length; i ++) {
        //std::cout << (char)((buffer[i] >> 24) & 0xFFFF) << (char)((buffer[i] >> 16) & 0xFFFF) << (char)((buffer[i] >> 8) & 0xFFFF) << (char)((buffer[i]) & 0xFFFF) << ", " << i << std::endl;
        if(!writer->writeChunk(buffer[i]))
            return false;
    }
    
    return true;
}

void Sharc::reset() {
    for(unsigned int i = 0; i < (1 << HASH_BITS); i++)
        *(unsigned int*)(dictionary + i) = 0;
}

int main(int argc, char *argv[]) {
	if(argc <= 1)
		exit(0);
    
    Chrono* chrono = new Chrono();
    std::cout << "Chrono initialized" << std::endl;
    
    Sharc* sharc = new Sharc();
    std::cout << "Sharc initialized" << std::endl;
    
    const unsigned int readBufferSize = 1 << 24;  // Maximum depending on ENTRY.offset length
    std::cout << "Allocating " << readBufferSize << " as buffer read" << std::endl;
	byte* readArray = new byte[readBufferSize];
	byte* writeArray = new byte[readBufferSize];
    std::cout << "Read / write arrays initialized" << std::endl;
    
    for(int i = 1; i < argc; i ++) {
        //unsigned int outBits = 0;
        unsigned int totalRead = 0;
        unsigned int totalWritten = 0;
        
		std::string inFileName = std::string(argv[i]);
		std::string outFileName = inFileName + ".sha";
		
        //std::ifstream inFile (inFileName, std::ios::in | std::ios::binary);
		//std::ofstream outFile (outFileName, std::ios::out | std::ios::binary);
        
        FILE* inFile = fopen(inFileName.c_str(), "rb");
        FILE* outFile = fopen(outFileName.c_str(), "wb+");
        
		SubsWriter* swriter = new SubsWriter(writeArray/*readArray*/, readBufferSize);
        std::cout << "SubsWriter initialized" << std::endl;
        
		SharcWriter* writer = new SharcWriter(readArray/*writeArray*/, readBufferSize);
        std::cout << "SharcWriter initialized" << std::endl;
        
        /*writer->writeChunk((unsigned int)65266236);
        writer->flush();
        std::cout << "ok " << writer->getPosition() << std::endl;*/
        unsigned int bytesRead;
        
        chrono->start();
		while((bytesRead = (unsigned int)fread(readArray, sizeof(byte), readBufferSize, inFile)) > 0) {
			//inFile.read ((char*)(readArray), readBufferSize);
			//unsigned int bytesRead = (unsigned int);//(unsigned int)inFile.gcount();
			totalRead += bytesRead;
            
            swriter->setLimit(bytesRead/*writer->getPosition()*/);
            
            if(sharc->subs(readArray, swriter)) {
                //std::cout << "Subs : compress" << std::endl;
                swriter->flush();
            } else {
                std::cout << "Subs : no compress" << std::endl;
            }
            //std::cout << bytesRead/*writer->getPosition()*/ << " > " << swriter->getPosition() << std::endl;
            
            writer->setLimit(swriter->getPosition()/*bytesRead*/);
             
             if(sharc->compress(readArray, writer)) {
                 writer->flush(); // todo manage remaining bytes
                 totalWritten += fwrite(readArray, sizeof(byte), writer->getPosition(), outFile);
             } else {
                 std::cout << "Sharc : No compress" << std::endl;
                 totalWritten += fwrite(writeArray, sizeof(byte), bytesRead, outFile);
             }
             //std::cout << swriter->getPosition()/*bytesRead*/ << " > " << writer->getPosition() << std::endl << "-----" << std::endl;
            
            
            sharc->reset();
            swriter->resetModes();
            swriter->resetBuffer();
            writer->resetModes();
            writer->resetBuffer();
		}
		chrono->stop();
                
        //outFile.flush();

		std::cout << "--------------------------------------------------------------------" << std::endl;
        
        double outBytes = totalWritten;
		double ratio = outBytes / totalRead;
		std::cout << "File " << argv[i] << ", " << totalRead << " bytes in, " << (unsigned int)outBytes << " bytes out" << std::endl;
        
		double outSpeed = (1000.0 * totalRead) / (chrono->getElapsedMillis() * 1024.0 * 1024.0);
		std::cout  << "Ratio out / in = " << ratio << ", time = " << chrono->getElapsedMillis() << " ms, Speed = " << outSpeed << " MB/s" << std::endl;
        
		std::cout << "COMBINED SCORE = " << outSpeed / ratio << std::endl << "------" << std::endl;
        
        fclose(inFile);
        fclose(outFile);
		//inFile.close();
		//outFile.close();
		//delete bitOutputStream;
		//delete inFile;
		//delete outFile;
    }
    
	delete readArray;
	delete sharc;
    delete chrono;
}
